AttributeError: 'Namespace' object has no attribute 'run_requires'

- [x] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Steps to reproduce

Same as #293 but I'll try to write up a test case to reproduce this more easily.

<!--Describe the minimized example of how to reproduce the bug-->

## Actual behavior

With 1.4.1 it now says:

```
uhid_freebsd.cpp:15:10: fatal error: 'dev/usb/usb_ioctl.h' file not found
#include <dev/usb/usb_ioctl.h>
         ^~~~~~~~~~~~~~~~~~~~~
1 error generated.
error: command '/usr/bin/clang' failed with exit code 1
Failed to build package, try parsing project files.
Traceback (most recent call last):
  File "/usr/local/bin/pdm", line 8, in <module>
    sys.exit(main())
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/core.py", line 76, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/core.py", line 131, in main
    raise err.with_traceback(traceback)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/core.py", line 127, in main
    f(options.project, options)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/cli/commands/install.py", line 33, in handle
    actions.do_lock(project, strategy="all")
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/cli/actions.py", line 70, in do_lock
    mapping, dependencies, summaries = resolve(
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/resolver/core.py", line 152, in resolve
    result = resolver.resolve(requirements, max_rounds)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/resolvelib/resolvers.py", line 454, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/resolvelib/resolvers.py", line 348, in resolve
    failure_causes = self._attempt_to_pin_criterion(name, criterion)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/resolvelib/resolvers.py", line 208, in _attempt_to_pin_criterion
    criteria = self._get_criteria_to_update(candidate)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/resolvelib/resolvers.py", line 199, in _get_criteria_to_update
    for r in self._p.get_dependencies(candidate=candidate):
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/resolver/providers.py", line 77, in get_dependencies
    deps, requires_python, summary = self.repository.get_dependencies(candidate)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/models/repositories.py", line 64, in get_dependencies
    requirements, requires_python, summary = getter(candidate)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/models/repositories.py", line 30, in wrapper
    result = func(self, candidate)
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/models/repositories.py", line 141, in _get_dependencies_from_metadata
    deps = candidate.get_dependencies_from_metadata()
  File "/usr/local/Cellar/pdm/1.4.1/libexec/lib/python3.9/site-packages/pdm/models/candidates.py", line 214, in get_dependencies_from_metadata
    return filter_requirements_with_extras(metadata.run_requires, extras)
AttributeError: 'Namespace' object has no attribute 'run_requires'
```

## Expected behavior

It installs `fido2==0.9.1` on non-FreeBSD systems.

## Environment Information

```bash
# Paste the output of `pdm info && pdm info --env` below:
PDM version:        1.4.1                                                   
Python Interpreter: .../.venv/bin/python (3.9.2)
Project Root:      ...                         
{
  "implementation_name": "cpython",
  "implementation_version": "3.9.2",
  "os_name": "posix",
  "platform_machine": "x86_64",
  "platform_release": "20.4.0",
  "platform_system": "Darwin",
  "platform_version": "Darwin Kernel Version 20.4.0: ...",
  "python_full_version": "3.9.2",
  "platform_python_implementaiton": "CPython",
  "python_version": "3.9",
  "sys_platform": "darwin"
}
```

--- a/news/299.bugfix.md
+++ b/news/299.bugfix.md
@@ -1,0 +1,1 @@
+Fix a bug that get_dependencies() returns error when the `setup.py` has no `intall_requires` key.

--- a/pdm/models/candidates.py
+++ b/pdm/models/candidates.py
@@ -205,8 +205,8 @@ class Candidate:
             if not metadata:
                 return []
             return get_requirements_from_dist(self.ireq.get_dist(), extras)
-        elif getattr(metadata, "install_requires", None):
-            requires = metadata.install_requires
+        elif hasattr(metadata, "install_requires"):
+            requires = metadata.install_requires or []
             for extra in extras:
                 requires.extend(metadata.extras_require.get(extra, []))
             return sorted(set(requires))

--- a/pdm/models/candidates.py
+++ b/pdm/models/candidates.py
@@ -207,8 +207,14 @@ class Candidate:
             return get_requirements_from_dist(self.ireq.get_dist(), extras)
         elif hasattr(metadata, "install_requires"):
             requires = metadata.install_requires or []
+            extras_not_found = set()
             for extra in extras:
-                requires.extend(metadata.extras_require.get(extra, []))
+                try:
+                    requires.extend((metadata.extras_require or {})[extra])
+                except KeyError:
+                    extras_not_found.add(extra)
+            if extras_not_found:
+                warnings.warn(ExtrasError(sorted(extras_not_found)))
             return sorted(set(requires))
         else:
             return filter_requirements_with_extras(metadata.run_requires, extras)

--- a/pdm/models/requirements.py
+++ b/pdm/models/requirements.py
@@ -435,7 +435,7 @@ def filter_requirements_with_extras(
 
     extras_not_found = [e for e in extras if e not in extras_in_meta]
     if extras_not_found:
-        warnings.warn(ExtrasError(extras_not_found), stacklevel=2)
+        warnings.warn(ExtrasError(extras_not_found))
 
     return result
 

--- a/tests/fixtures/projects/demo-failure-no-dep/demo.py
+++ b/tests/fixtures/projects/demo-failure-no-dep/demo.py
@@ -1,0 +1,5 @@
+import os
+
+import chardet
+
+print(os.name)

--- a/tests/fixtures/projects/demo-failure-no-dep/setup.py
+++ b/tests/fixtures/projects/demo-failure-no-dep/setup.py
@@ -1,0 +1,12 @@
+from setuptools import setup
+
+if True:
+    raise RuntimeError("This mimics the build error on unmatched platform")
+
+setup(
+    name="demo",
+    version="0.0.1",
+    description="test demo",
+    py_modules=["demo"],
+    python_requires=">=3.3",
+)

--- a/tests/models/test_candidates.py
+++ b/tests/models/test_candidates.py
@@ -133,3 +133,11 @@ def test_parse_project_file_on_build_error(project):
     ]
     assert candidate.name == "demo"
     assert candidate.version == "0.0.1"
+
+
+def test_parse_project_file_on_build_error_no_dep(project):
+    req = parse_requirement(f"{(FIXTURES / 'projects/demo-failure-no-dep').as_posix()}")
+    candidate = Candidate(req, project.environment)
+    assert candidate.get_dependencies_from_metadata() == []
+    assert candidate.name == "demo"
+    assert candidate.version == "0.0.1"

